ReactAndroid/src/main/java/com/facebook/react/modules/location/LocationModule.java

```diff
@@ -188,18 +190,34 @@ public void stopObserving() {
     mWatchedProvider = null;
   }
 
+
+  private boolean isProviderEnabled(LocationManager locationManager, String provider) {
+    try {
+      // Before API version LOLLIPOP, this method would throw SecurityException
+      // if the location permissions were not sufficient to use the specified provider.
+      return locationManager.isProviderEnabled(provider);
+    } catch (SecurityException e) {
+      return false;
+    }
+  }
+
   @Nullable
-  private static String getValidProvider(LocationManager locationManager, boolean highAccuracy) {
+  private String getValidProvider(LocationManager locationManager, boolean highAccuracy) {
     String provider =
         highAccuracy ? LocationManager.GPS_PROVIDER : LocationManager.NETWORK_PROVIDER;
-    if (!locationManager.isProviderEnabled(provider)) {
+    if (!isProviderEnabled(locationManager, provider)) {
       provider = provider.equals(LocationManager.GPS_PROVIDER)
           ? LocationManager.NETWORK_PROVIDER
           : LocationManager.GPS_PROVIDER;
-      if (!locationManager.isProviderEnabled(provider)) {
+      if (!isProviderEnabled(locationManager, provider)) {
         return null;
       }
     }
+    // If it's an enabled provider, but we don't have permissions, ignore it
+    int finePermission = ContextCompat.checkSelfPermission(getReactApplicationContext(), android.Manifest.permission.ACCESS_FINE_LOCATION);
+    if (provider.equals(LocationManager.GPS_PROVIDER) && finePermission != PackageManager.PERMISSION_GRANTED) {
+      return null;
```

Comment:
The existing error cases above `return null`, so I was just following that example. I think that's all `getValidProvider` can do.

It's called from two places.
- When the `getCurrentPosition` call receives null, it will call `error.invoke("No available location provider.");`
- When the `startObserving` call receives null, it will emit `geolocationError` and `watchPosition()`'s error function.